### PR TITLE
feat: make save and save as work with knowledge

### PR DESCRIPTION
**Pull request by thedadams**

### CODE REVIEW

File: components/scripts/knowledge-dropdown.tsx [hunk 1]
```diff
@@ -66,40 +99,114 @@ const ScriptKnowledgeDropdown = () => {
...
         >
-          {knowledgeFiles && knowledgeFiles.length ? (
-            <DropdownSection
-              title={
-                'Thread Knowledge' +
-                (knowledgeFiles.length > 0 ? ' (Click to remove)' : '')
-              }
-            >
-              {knowledgeFiles.map((f) => (
+          <DropdownSection
```
Comment (thedadams) [inline components/scripts/knowledge-dropdown.tsx:107]: Most of this is getting the counts displayed correctly and adding the "show all" functionality when there is more than 10 files.

File: components/chat/useChatSocket.tsx [hunk 2]
```diff
@@ -264,8 +267,27 @@ const useChatSocket = (isEmpty?: boolean) => {
     []
   );
 
-  const handleToolAdjusted = (tools: string[]) => {
+  const handleToolChange = (
```
Comment (StrongMonkey) [inline components/chat/useChatSocket.tsx:270]: I will probably make this an async function. .then syntax is too hard to use and read.

File: components/scripts/knowledge-dropdown.tsx [hunk 3]
```diff
@@ -12,20 +12,48 @@ import { PiBookOpenThin, PiXThin } from 'react-icons/pi';
...
+  const [threadKnowledgeFiles, setThreadKnowledgeFiles] = useState<string[]>(
+    []
+  );
+  const [assistantKnowledgeFiles, setAssistantKnowledgeFiles] = useState<
+    string[]
+  >([]);
+  const [assistantExpanded, setAssistantExpanded] = useState(false);
+  const [threadExpanded, setThreadExpanded] = useState(false);
 
   function refreshKnowledgeFiles(isOpen: boolean) {
```
Comment (StrongMonkey) [inline components/scripts/knowledge-dropdown.tsx:38]: Same here..make whole async if possible

File: components/scripts/knowledge-dropdown.tsx (hunk 1 above)
Comment (StrongMonkey) [reply components/scripts/knowledge-dropdown.tsx:107]: I don't know if we should add max height and make it scrollable. Show like 100 files will still be too long if user clicked show all,

**Review by StrongMonkey — APPROVED** [review]

File: components/scripts/knowledge-dropdown.tsx (hunk 1 above)
Comment (thedadams) [reply components/scripts/knowledge-dropdown.tsx:107]: Not sure what you are suggesting here. The max height and scroll works for lots of files.

Comment (StrongMonkey) [reply components/scripts/knowledge-dropdown.tsx:107]: I am suggesting adding a max height and make it scrollable. Show all will still mess up with dropdown if there are a lot of files.

Comment (thedadams) [reply components/scripts/knowledge-dropdown.tsx:107]: Oh, cool. That is already here. So we're good!

Comment (thedadams) [reply components/scripts/knowledge-dropdown.tsx:107]: Here is an example.

Screenshot 2024-08-30 at 07 43 36

Screenshot 2024-08-30 at 07 43 59

After I took those screenshots, I made the "Show All" text smaller.
Screenshot 2024-08-30 at 07 47 56

File: components/chat/useChatSocket.tsx (hunk 2 above)
Comment (thedadams) [reply components/chat/useChatSocket.tsx:270]: Done.

File: components/scripts/knowledge-dropdown.tsx (hunk 3 above)
Comment (thedadams) [reply components/scripts/knowledge-dropdown.tsx]: Done.

**Review by g-linville — APPROVED** [review]

**Pull request merged**